### DISCUSSION TITLE
Add a prefix and a suffix to a string

### DISCUSSION THREAD

**Discussion by NikosAlexandris** [johnkerl/miller#647] [Q&A] [answered]
How can one simply add a prefix and a suffix to a string, using `put` and `gsub` or else `ssub`?

I am trying something like
```
put -S '$v=ssub($v,"^([^_].*)","_$v")'
```
for adding a `_` as a prefix, but it obviously does not work.

**Comment by NikosAlexandris** [comment]
The following works:
```
then put -S '$v=sub($v,"^","`")' then put -S '$v=sub($v,"$","`")'
```

But, I can't seem to make a single call out of it... 🤷🏼

**Reply by trantor** [reply]
Mmm @NikosAlexandris is your question about adding a prefix/suffix in general or about how to do that using `sub`/`gsub` in a `put` command?

Because you can always do
```
$ echo 1 | mlr -N --csv put '$1 = "PREF_".$1."_SUFF"'
PREF_1_SUFF
```
but the question is about `sub`, for instance, look out for the data type
```
$ echo 1 | mlr -N --csv put ' $1=sub(string($1),"^","PREF_")'
PREF_1
```
Here I've used `string` to cast `$1` as one.

**Reply by NikosAlexandris** [reply]
Indeed, I am interested in an `mlr` solution using [`sub`](https://johnkerl.org/miller6/reference-dsl-builtin-functions/#sub) or [`gsub`](https://johnkerl.org/miller6/reference-dsl-builtin-functions/#gsub) or [`ssub`](https://johnkerl.org/miller6/reference-dsl-builtin-functions/#ssub).

In this case all input data are strings -- but, thank you, I will keep it in mind to check for the data type.
And, why not including the `_SUFF` in the `sub` example? I.e. how to do prefix and suffix in one step using `sub`?

**Reply by aborruso** [reply]
> And, why not including the `_SUFF` in the `sub` example? I.e. how to do prefix and suffix in one step using `sub`?

What about using regex syntax 

Running `echo 'a=hello' | mlr put -S '$a=sub($a,"^(.+)$","PRE_"."\1"."_SUFF")'` you will have `a=PRE_hello_SUFF`

**Reply by NikosAlexandris** [reply]
Thanks @aborruso -- I was looking for this! `put -S '$v=sub($v,"^(.+)$","`"."\1"."`")'` (for me and also for https://github.com/aborruso/millerQuestions/pull/2/commits/8a4a47fe4995c70f69fdc3029a2a18ab887101be#diff-27033d9dd92fa04ad7a610845e3354c61fc8ce5fe2abd7c35c17c260cd16e2adR23).

**Reply by NikosAlexandris** [reply]
> > And, why not including the `_SUFF` in the `sub` example? I.e. how to do prefix and suffix in one step using `sub`?
> 
> What about using regex syntax
> 
> Running `echo 'a=hello' | mlr put -S '$a=sub($a,"^(.+)$","PRE_"."\1"."_SUFF")'` you will have `a=PRE_hello_SUFF`

Can you please put it in a separate answer, so I can mark it as answered?

**Reply by trantor** [reply]
> In this case all input data are strings -- but, thank you, I will keep it in mind to check for the data type.
> And, why not including the `_SUFF` in the `sub` example? I.e. how to do prefix and suffix in one step using `sub`?

Well @NikosAlexandris , I was merely referring to your example using the start of pattern anchor. I thought that you might have had some additional trouble since, as @aborruso showed, the string substitution is rather similar to your average `awk` or similar, with `\0` `\1` and the likes as back-references.

~~There's an example of the back-reference expansion [here](https://miller.readthedocs.io/en/latest/reference-dsl.html#pattern-action-blocks) in passing, however the whole regex topic is not exactly explained in detail in the docs. That might be something to improve upon for the new release @johnkerl .~~
My bad, [here it is](https://miller.readthedocs.io/en/latest/reference.html#regex-captures)

Anyway @aborruso a (imho) better answer in this case is
```
$ echo 'a=hello' | mlr put -S '$a=sub($a,".*","PRE_\0_SUFF")'
a=PRE_hello_SUFF
```
No needless concatenation operators, no unnecessary capture group.

**Comment by aborruso** [comment, marked as answer]
> And, why not including the `_SUFF` in the `sub` example? I.e. how to do prefix and suffix in one step using `sub`?

What about using regex syntax 

Running `echo 'a=hello' | mlr put -S '$a=sub($a,"^(.+)$","PRE_"."\1"."_SUFF")'` you will have `a=PRE_hello_SUFF`

**Comment by johnkerl** [comment]
@NikosAlexandris can you show some sample input and desired output? At present I think you can do `$v = "prefix" . $v . "suffix"` but maybe I misunderstand the question ...

**Reply by NikosAlexandris** [reply]
@johnkerl This is happening here: https://github.com/aborruso/millerQuestions/blob/8b77f635d6a25191342892dcc141c48dcdc9e07b/script/millerQuestions.sh#L23 with input from https://github.com/aborruso/millerQuestions/blob/8b77f635d6a25191342892dcc141c48dcdc9e07b/data/millerQuestions.yml#L1-L5 for example.

**Reply by NikosAlexandris** [reply]
Thank you @johnkerl! Way easier and readable... Will update the script.
